system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -20,105 +37,171 @@ std::string normalize(const std::string& name) {
+std::shared_ptr<Directory>
+IgnoreCaseDirectory::insertDirectory(const std::string& name, mode_t mode) {
+  auto normalized = normalize(name);
+  auto dir = real->locked().insertDirectory(normalized, mode);
+  if (!dir) {
+    return nullptr;
   }
+  dir = virtualize(dir, getBackend());
+  children[normalized] = {name, dir};
+  return dir;
+}
 
-  int removeChild(const std::string& name) override {
-    auto name2 = normalize(name);
-    if (auto err = BaseClass::removeChild(name2))
-      return err;
-    return baseDirectory->locked().removeChild(name2);
+std::shared_ptr<Symlink>
+IgnoreCaseDirectory::insertSymlink(const std::string& name,
+                                   const std::string& target) {
+  auto normalized = normalize(name);
+  auto link = real->locked().insertSymlink(normalized, target);
+  if (!link) {
+    return nullptr;
   }
+  link = virtualize(link, getBackend());
+  children[normalized] = {name, link};
+  return link;
+}
 
-  ssize_t getNumEntries() override { return baseDirectory->locked().getNumEntries(); }
+int IgnoreCaseDirectory::insertMove(const std::string& name,
+                                    std::shared_ptr<File> file) {
+  auto normalized = normalize(name);
+  if (auto err = real->locked().insertMove(normalized, devirtualize(file))) {
+    return err;
+  }
+  auto oldParent =
+    std::static_pointer_cast<IgnoreCaseDirectory>(file->locked().getParent());
+  auto& oldChildren = oldParent->children;
+  for (auto it = oldChildren.begin(); it != oldChildren.end(); ++it) {
+    if (it->second.child == file) {
+      oldChildren.erase(it);
+      break;
+    }
+  }
+  children[normalized] = {name, file};
+  return 0;
+}
 
-  // TODO: preserve original name, denormalize, and use it here
-  Directory::MaybeEntries getEntries() override {
-    return baseDirectory->locked().getEntries();
+int IgnoreCaseDirectory::removeChild(const std::string& name) {
+  auto normalized = normalize(name);
+  if (auto err = real->locked().removeChild(normalized)) {
+    return err;
   }
+  auto it = children.find(normalized);
+  assert(it != children.end());
+  it->second.child->locked().setParent(nullptr);
+  children.erase(it);
+  return 0;
+}
 
-  // TODO: preserve original name, denormalize, and use it here
-  std::string getName(std::shared_ptr<File> file) override {
-    return BaseClass::getName(file);
+Directory::MaybeEntries IgnoreCaseDirectory::getEntries() {
+  auto entries = real->locked().getEntries();
+  if (auto err = entries.getError()) {
+    return entries;
   }
+  for (auto& entry : *entries) {
+    if (auto it = children.find(entry.name); it != children.end()) {
+      entry.name = it->second.originalName;
+    }
+  }
+  return entries;
+}
 
-  bool maintainsFileIdentity() override { return true; }
-};
+std::string IgnoreCaseDirectory::getName(std::shared_ptr<File> file) {
+  for (auto& [_, info] : children) {
+    if (info.child == file) {
+      return info.originalName;
+    }
+  }
+  return "";
```

Comment:
Maybe, fallback to `real->getName()` here?

Reply:
No, that wouldn't be correct because we must be getting the name of a virtual `File` here, not one of the underlying "real" files. That's because the underlying real `File` objects are visible only to the virtual backend and never escape into the rest of the WasmFS system.